File: projects/dejavu.ts/dejavu-engine/src/render.ts
import {
    applyFilter,
    htmlEscape,
    isSafeHtml,
    valueToString,
    type CanonicalId,
    type IrDocument,
    type IrNode,
    type IrValue,
    type RenderOptions,
    type TemplateResolveOk,
    type TemplateLoader,
} from "@dejavu/types";

function loadTemplate(
    loader: TemplateLoader,
    ref: string,
    from: CanonicalId,
    onDependency?: (id: CanonicalId) => void,
): TemplateResolveOk {
    const loaded = loader.resolve(ref, { from });
    onDependency?.(loaded.id);
    return loaded;
}

function noteDependency(env: RenderEnv, id: CanonicalId): void {
    env.onDependency?.(id);
}

type BlockNode = Extract<IrNode, { type: "Stmt.Block" }> & {
    __parentDefault?: IrNode[];
};

function truthy(v: IrValue): boolean {
    if (v === null) return false;
    if (typeof v === "boolean") return v;
    if (typeof v === "number") return v !== 0;
    if (typeof v === "string") return v.length > 0;
    if (Array.isArray(v)) return v.length > 0;
    if (typeof v === "object") return Object.keys(v).length > 0;
    return false;
}

/** Render IR with a JSON object context. Optional loader enables extends/include/super. */
export function renderIr(
    doc: IrDocument,
    ctx: Record<string, IrValue> = {},
    options: RenderOptions = {},
): string {
    const scope = { ...ctx };
    let name: CanonicalId = options.name ?? "<main>";
    // Canonicalize entry name when possible so relative includes get a valid `from`.
    if (options.loader && name !== "<main>" && !/^[A-Za-z][A-Za-z0-9_+-]*:/.test(name)) {
        try {
            name = options.loader.resolve(name).id;
        } catch {
            /* keep caller name for T1 / ad-hoc docs */
        }
    }
    const env: RenderEnv = {
        loader: options.loader,
        stack: [name],
        parentBlocks: [],
        strictUndefined: options.strictUndefined === true,
        onDependency: options.onDependency,
    };
    if (name !== "<main>") {
        noteDependency(env, name);
    }
    const resolved = resolveInheritance(doc, scope, env, name, new Set());
    return renderNode(resolved.body, scope, env);
}

type RenderEnv = {
    loader?: TemplateLoader;
    stack: CanonicalId[];
    parentBlocks: IrNode[][];
    strictUndefined: boolean;
    onDependency?: (id: CanonicalId) => void;
};

function templateChildren(doc: IrDocument): IrNode[] {
    if (doc.body.type !== "Template") {
        throw new Error("IrDocument.body must be Template");
    }
    return doc.body.children;
}

function collectBlocks(nodes: IrNode[]): Map<string, IrNode[]> {
    const map = new Map<string, IrNode[]>();
    for (const n of nodes) {
        if (n.type === "Stmt.Block") map.set(n.name, n.body);
    }
    return map;
}

function findExtends(nodes: IrNode[]): Extract<IrNode, { type: "Stmt.Extends" }> | null {
    for (const n of nodes) {
        if (n.type === "Stmt.Extends") return n;
    }
    return null;
}

function requireStringPath(v: IrValue, kind: string): string {
    if (typeof v !== "string" || v.length === 0) {
        throw new Error(`${kind} path must evaluate to a non-empty string`);
    }
    return v;
}

function resolveInheritance(
    doc: IrDocument,
    scope: Record<string, IrValue>,
    env: RenderEnv,
    name: string,
    visiting: Set<string>,
): IrDocument {
    const children = templateChildren(doc);
    const ext = findExtends(children);
    if (!ext) return doc;
    if (!env.loader) {
        throw new Error(
            "extends/include/super require a template loader (not in T1 minimal render)",
        );
    }
    if (visiting.has(name)) {
        throw new Error(`circular template inheritance involving '${name}'`);
    }
    visiting.add(name);

    const parentRef = requireStringPath(evalExpr(ext.parent, scope, env), "extends");
    const parent = loadTemplate(env.loader, parentRef, name, env.onDependency);
    if (visiting.has(parent.id)) {
        throw new Error(`circular template inheritance: ${[...visiting, parent.id].join(" -> ")}`);
    }
    const parentResolved = resolveInheritance(parent.document, scope, env, parent.id, visiting);
    visiting.delete(name);

    const childBlocks = collectBlocks(children);
    const mergedChildren = applyBlocks(templateChildren(parentResolved), childBlocks);
    return {
        ...parentResolved,
        body: { type: "Template", children: mergedChildren },
    };
}

function applyBlocks(nodes: IrNode[], overrides: Map<string, IrNode[]>): IrNode[] {
    return nodes.map((n) => {
        if (n.type === "Stmt.Block") {
            const override = overrides.get(n.name);
            if (!override) {
                return { ...n, body: applyBlocks(n.body, overrides) };
            }
            const parentDefault = applyBlocks(n.body, overrides);
            const merged: BlockNode = {
                type: "Stmt.Block",
                name: n.name,
                trim: n.trim,
                body: override,
                __parentDefault: parentDefault,
            };
            return merged;
        }
        if (n.type === "Template") {
            return { ...n, children: applyBlocks(n.children, overrides) };
        }
        if (n.type === "Stmt.If") {
            return {
                ...n,
                consequent: applyBlocks(n.consequent, overrides),
                elseIfs: n.elseIfs.map((ei) =>
                    ei.type === "Stmt.ElseIf"
                        ? { ...ei, consequent: applyBlocks(ei.consequent, overrides) }
                        : ei,
                ),
                ...(n.alternate ? { alternate: applyBlocks(n.alternate, overrides) } : {}),
            };
        }
        if (n.type === "Stmt.For") {
            return { ...n, body: applyBlocks(n.body, overrides) };
        }
        return n;
    });
}

function renderNode(node: IrNode, scope: Record<string, IrValue>, env: RenderEnv): string {
    switch (node.type) {
        case "Template":
            return node.children.map((c) => renderNode(c, scope, env)).join("");
        case "Text":
            return node.value;
        case "Comment":
            return "";
        case "Interpolation": {
            const v = evalExpr(node.expression, scope, env);
            const s = valueToString(v);
            if (node.raw || isSafeHtml(v)) return s;
            return htmlEscape(s);
        }
        case "Stmt.If": {
            if (truthy(evalExpr(node.test, scope, env))) {
                return node.consequent.map((c) => renderNode(c, scope, env)).join("");
            }
            for (const ei of node.elseIfs) {
                if (ei.type === "Stmt.ElseIf" && truthy(evalExpr(ei.test, scope, env))) {
                    return ei.consequent.map((c) => renderNode(c, scope, env)).join("");
                }
            }
            if (node.alternate) {
                return node.alternate.map((c) => renderNode(c, scope, env)).join("");
            }
            return "";
        }
        case "Stmt.For": {
            const iter = evalExpr(node.iterable, scope, env);
            if (!Array.isArray(iter)) {
                throw new Error("for iterable must be array");
            }
            let out = "";
            iter.forEach((val, i) => {
                const prev = scope[node.item];
                scope[node.item] = val;
                let prevIdx: IrValue | undefined;
                if (node.index) {
                    prevIdx = scope[node.index];
                    scope[node.index] = i;
                }
                out += node.body.map((c) => renderNode(c, scope, env)).join("");
                if (prev === undefined) delete scope[node.item];
                else scope[node.item] = prev;
                if (node.index) {
                    if (prevIdx === undefined) delete scope[node.index];
                    else scope[node.index] = prevIdx;
                }
            });
            return out;
        }
        case "Stmt.Raw":
            return node.value;
        case "Stmt.Block": {
            const block = node as BlockNode;
            const nextEnv: RenderEnv = block.__parentDefault
                ? { ...env, parentBlocks: [...env.parentBlocks, block.__parentDefault] }
                : env;
            return block.body.map((c) => renderNode(c, scope, nextEnv)).join("");
        }
        case "Stmt.Super": {
            const parent = env.parentBlocks[env.parentBlocks.length - 1];
            if (!parent) {
                throw new Error("`super` used outside an overriding block");
            }
            const nextEnv: RenderEnv = {
                ...env,
                parentBlocks: env.parentBlocks.slice(0, -1),
            };
            return parent.map((c) => renderNode(c, scope, nextEnv)).join("");
        }
        case "Stmt.Extends":
            return "";
        case "Stmt.Include": {
            if (!env.loader) {
                throw new Error(
                    "extends/include/super require a template loader (not in T1 minimal render)",
                );
            }
            const includeRef = requireStringPath(evalExpr(node.path, scope, env), "include");
            const from = env.stack[env.stack.length - 1] ?? "<main>";
            const included = loadTemplate(env.loader, includeRef, from, env.onDependency);
            if (env.stack.includes(included.id)) {
                throw new Error(
                    `circular template include: ${[...env.stack, included.id].join(" -> ")}`,
                );
            }
            const resolved = resolveInheritance(
                included.document,
                scope,
                env,
                included.id,
                new Set(env.stack),
            );
            return renderNode(resolved.body, scope, {
                ...env,
                stack: [...env.stack, included.id],
            });
        }
        default:
            throw new Error(`node not renderable: ${(node as IrNode).type}`);
    }
}

function evalExpr(expr: IrNode, scope: Record<string, IrValue>, env: RenderEnv): IrValue {
    switch (expr.type) {
        case "Expr.Literal":
            return expr.value;
        case "Expr.Identifier": {
            if (!(expr.name in scope)) {
                if (env.strictUndefined) {
                    throw new Error(`undefined variable \`${expr.name}\``);
                }
                return null;
            }
            return scope[expr.name] ?? null;
        }
        case "Expr.Member": {
            const obj = evalExpr(expr.object, scope, env);
            if (obj && typeof obj === "object" && !Array.isArray(obj) && !isSafeHtml(obj)) {
                return (obj as Record<string, IrValue>)[expr.property] ?? null;
            }
            return null;
        }
        case "Expr.Index": {
            const obj = evalExpr(expr.object, scope, env);
            const idx = evalExpr(expr.index, scope, env);
            if (Array.isArray(obj) && typeof idx === "number") {
                return obj[idx] ?? null;
            }
            if (obj && typeof obj === "object" && !isSafeHtml(obj) && typeof idx === "string") {
                return (obj as Record<string, IrValue>)[idx] ?? null;
            }
            return null;
        }
        case "Expr.Binary": {
            const l = evalExpr(expr.left, scope, env);
            const r = evalExpr(expr.right, scope, env);
            return evalBinary(expr.operator, l, r);
        }
        case "Expr.Unary": {
            const v = evalExpr(expr.argument, scope, env);
            if (expr.operator === "!") return !truthy(v);
            if (expr.operator === "-" && typeof v === "number") return -v;
            if (expr.operator === "+") return v;
            return null;
        }
        case "Expr.Pipe": {
            let val = evalExpr(expr.expression, scope, env);
            const args = expr.arguments.map((a) => evalExpr(a, scope, env));
            val = applyFilter(expr.filter, val, args);
            return val;
        }
        case "Expr.Call":
            throw new Error("calls not supported in T1 eval");
        default:
            throw new Error(`invalid expression: ${(expr as IrNode).type}`);
    }
}

function evalBinary(op: string, l: IrValue, r: IrValue): IrValue {
    switch (op) {
        case "+":
            if (typeof l === "number" && typeof r === "number") return l + r;
            return valueToString(l) + valueToString(r);
        case "-":
            return typeof l === "number" && typeof r === "number" ? l - r : null;
        case "*":
            return typeof l === "number" && typeof r === "number" ? l * r : null;
        case "/":
            return typeof l === "number" && typeof r === "number" ? l / r : null;
        case "%":
            return typeof l === "number" && typeof r === "number" ? l % r : null;
        case "==":
            return l === r;
        case "!=":
            return l !== r;
        case "<":
            return typeof l === "number" && typeof r === "number" ? l < r : false;
        case "<=":
            return typeof l === "number" && typeof r === "number" ? l <= r : false;
        case ">":
            return typeof l === "number" && typeof r === "number" ? l > r : false;
        case ">=":
            return typeof l === "number" && typeof r === "number" ? l >= r : false;
        case "&&":
            return truthy(l) && truthy(r);
        case "||":
            return truthy(l) || truthy(r);
        case "in":
            if (Array.isArray(r)) return r.includes(l);
            if (typeof r === "string") return r.includes(valueToString(l));
            return false;
        default:
            throw new Error(`unknown binary ${op}`);
    }
}

File: projects/dejavu.rs/dejavu-ir/src/render.rs
use crate::IrError;
use crate::types::{IrDocument, IrNode};
use serde_json::Value;
use std::collections::HashMap;

/// Render an IR document with a JSON object context.
pub fn render_ir(doc: &IrDocument, ctx: &Value) -> Result<String, IrError> {
    let mut scope = json_object_to_map(ctx)?;
    render_nodes(&doc.body, &mut scope)
}

fn json_object_to_map(ctx: &Value) -> Result<HashMap<String, Value>, IrError> {
    match ctx {
        Value::Object(map) => Ok(map.iter().map(|(k, v)| (k.clone(), v.clone())).collect()),
        _ => Err(IrError::Render("context must be a JSON object".into())),
    }
}

fn render_nodes(node: &IrNode, scope: &mut HashMap<String, Value>) -> Result<String, IrError> {
    match node {
        IrNode::Template { children, .. } => {
            let mut out = String::new();
            for c in children {
                out.push_str(&render_nodes(c, scope)?);
            }
            Ok(out)
        }
        IrNode::Text { value, .. } => Ok(value.clone()),
        IrNode::Comment { .. } => Ok(String::new()),
        IrNode::Interpolation { expression, raw, .. } => {
            let v = eval(expression, scope)?;
            let s = value_to_string(&v);
            if *raw { Ok(s) } else { Ok(html_escape(&s)) }
        }
        IrNode::StmtIf { test, consequent, else_ifs, alternate, .. } => {
            if is_truthy(&eval(test, scope)?) {
                return render_list(consequent, scope);
            }
            for ei in else_ifs {
                if let IrNode::StmtElseIf { test, consequent, .. } = ei {
                    if is_truthy(&eval(test, scope)?) {
                        return render_list(consequent, scope);
                    }
                }
            }
            if let Some(alt) = alternate {
                return render_list(alt, scope);
            }
            Ok(String::new())
        }
        IrNode::StmtFor { item, index, iterable, body, .. } => {
            let iter_val = eval(iterable, scope)?;
            let items = match iter_val {
                Value::Array(a) => a,
                other => return Err(IrError::Render(format!("for iterable must be array, got {other}"))),
            };
            let mut out = String::new();
            for (i, val) in items.into_iter().enumerate() {
                let prev = scope.insert(item.clone(), val);
                let prev_idx = if let Some(idx) = index { scope.insert(idx.clone(), Value::Number(i.into())) } else { None };
                out.push_str(&render_list(body, scope)?);
                restore(scope, item, prev);
                if let Some(idx) = index {
                    restore(scope, idx, prev_idx);
                }
            }
            Ok(out)
        }
        IrNode::StmtRaw { value, .. } => Ok(value.clone()),
        IrNode::StmtBlock { body, .. } => render_list(body, scope),
        IrNode::StmtExtends { .. } | IrNode::StmtInclude { .. } | IrNode::StmtSuper { .. } => {
            Err(IrError::Render("extends/include/super require a template loader (not in T1 minimal render)".into()))
        }
        other => Err(IrError::Render(format!("node not renderable at top level: {other:?}"))),
    }
}

fn render_list(nodes: &[IrNode], scope: &mut HashMap<String, Value>) -> Result<String, IrError> {
    let mut out = String::new();
    for n in nodes {
        out.push_str(&render_nodes(n, scope)?);
    }
    Ok(out)
}

fn restore(scope: &mut HashMap<String, Value>, key: &str, prev: Option<Value>) {
    match prev {
        Some(v) => {
            scope.insert(key.to_string(), v);
        }
        None => {
            scope.remove(key);
        }
    }
}

fn eval(expr: &IrNode, scope: &HashMap<String, Value>) -> Result<Value, IrError> {
    match expr {
        IrNode::ExprLiteral { value, .. } => Ok(value.clone()),
        IrNode::ExprIdentifier { name, .. } => Ok(scope.get(name).cloned().unwrap_or(Value::Null)),
        IrNode::ExprMember { object, property, .. } => {
            let obj = eval(object, scope)?;
            match obj {
                Value::Object(map) => Ok(map.get(property).cloned().unwrap_or(Value::Null)),
                _ => Ok(Value::Null),
            }
        }
        IrNode::ExprIndex { object, index, .. } => {
            let obj = eval(object, scope)?;
            let idx = eval(index, scope)?;
            match (obj, idx) {
                (Value::Array(a), Value::Number(n)) => {
                    let i = n.as_u64().unwrap_or(0) as usize;
                    Ok(a.get(i).cloned().unwrap_or(Value::Null))
                }
                (Value::Object(map), Value::String(k)) => Ok(map.get(&k).cloned().unwrap_or(Value::Null)),
                _ => Ok(Value::Null),
            }
        }
        IrNode::ExprCall { .. } => Err(IrError::Render("calls not supported in T1 eval".into())),
        IrNode::ExprBinary { operator, left, right, .. } => {
            let l = eval(left, scope)?;
            let r = eval(right, scope)?;
            eval_binary(operator, &l, &r)
        }
        IrNode::ExprUnary { operator, argument, .. } => {
            let v = eval(argument, scope)?;
            match operator.as_str() {
                "!" => Ok(Value::Bool(!is_truthy(&v))),
                "-" => match v {
                    Value::Number(n) => {
                        if let Some(i) = n.as_i64() {
                            Ok(Value::Number((-i).into()))
                        } else if let Some(f) = n.as_f64() {
                            Ok(serde_json::Number::from_f64(-f).map(Value::Number).unwrap_or(Value::Null))
                        } else {
                            Ok(Value::Null)
                        }
                    }
                    _ => Ok(Value::Null),
                },
                "+" => Ok(v),
                _ => Err(IrError::Render(format!("unknown unary {operator}"))),
            }
        }
        IrNode::ExprPipe { expression, filter, arguments, .. } => {
            let mut val = eval(expression, scope)?;
            let args: Result<Vec<_>, _> = arguments.iter().map(|a| eval(a, scope)).collect();
            val = apply_filter(filter, val, &args?)?;
            Ok(val)
        }
        _ => Err(IrError::Render("invalid expression node".into())),
    }
}

fn eval_binary(op: &str, l: &Value, r: &Value) -> Result<Value, IrError> {
    match op {
        "+" => match (l, r) {
            (Value::Number(a), Value::Number(b)) => Ok(num_op(a, b, |x, y| x + y)),
            _ => Ok(Value::String(format!("{}{}", value_to_string(l), value_to_string(r)))),
        },
        "-" => match (l, r) {
            (Value::Number(a), Value::Number(b)) => Ok(num_op(a, b, |x, y| x - y)),
            _ => Ok(Value::Null),
        },
        "*" => match (l, r) {
            (Value::Number(a), Value::Number(b)) => Ok(num_op(a, b, |x, y| x * y)),
            _ => Ok(Value::Null),
        },
        "/" => match (l, r) {
            (Value::Number(a), Value::Number(b)) => Ok(num_op(a, b, |x, y| x / y)),
            _ => Ok(Value::Null),
        },
        "%" => match (l, r) {
            (Value::Number(a), Value::Number(b)) => Ok(num_op(a, b, |x, y| x % y)),
            _ => Ok(Value::Null),
        },
        "==" => Ok(Value::Bool(l == r)),
        "!=" => Ok(Value::Bool(l != r)),
        "<" | "<=" | ">" | ">=" => match (as_f64(l), as_f64(r)) {
            (Some(a), Some(b)) => Ok(Value::Bool(match op {
                "<" => a < b,
                "<=" => a <= b,
                ">" => a > b,
                ">=" => a >= b,
                _ => false,
            })),
            _ => Ok(Value::Bool(false)),
        },
        "&&" => Ok(Value::Bool(is_truthy(l) && is_truthy(r))),
        "||" => Ok(Value::Bool(is_truthy(l) || is_truthy(r))),
        "in" => match r {
            Value::Array(a) => Ok(Value::Bool(a.contains(l))),
            Value::String(s) => Ok(Value::Bool(s.contains(&value_to_string(l)))),
            _ => Ok(Value::Bool(false)),
        },
        _ => Err(IrError::Render(format!("unknown binary {op}"))),
    }
}

fn num_op(a: &serde_json::Number, b: &serde_json::Number, f: impl Fn(f64, f64) -> f64) -> Value {
    match (as_f64(&Value::Number(a.clone())), as_f64(&Value::Number(b.clone()))) {
        (Some(x), Some(y)) => serde_json::Number::from_f64(f(x, y)).map(Value::Number).unwrap_or(Value::Null),
        _ => Value::Null,
    }
}

fn as_f64(v: &Value) -> Option<f64> {
    match v {
        Value::Number(n) => n.as_f64(),
        _ => None,
    }
}

fn apply_filter(name: &str, value: Value, args: &[Value]) -> Result<Value, IrError> {
    match name {
        "uppercase" => Ok(Value::String(value_to_string(&value).to_uppercase())),
        "lowercase" => Ok(Value::String(value_to_string(&value).to_lowercase())),
        "trim" => Ok(Value::String(value_to_string(&value).trim().to_string())),
        "default" => {
            if matches!(value, Value::Null) || matches!(&value, Value::String(s) if s.is_empty()) {
                Ok(args.first().cloned().unwrap_or(Value::Null))
            } else {
                Ok(value)
            }
        }
        "length" => match value {
            Value::String(s) => Ok(Value::Number(s.chars().count().into())),
            Value::Array(a) => Ok(Value::Number(a.len().into())),
            Value::Object(o) => Ok(Value::Number(o.len().into())),
            _ => Ok(Value::Number(0.into())),
        },
        "join" => {
            let sep = args.first().map(value_to_string).unwrap_or_else(|| ",".into());
            match value {
                Value::Array(a) => Ok(Value::String(a.iter().map(value_to_string).collect::<Vec<_>>().join(&sep))),
                _ => Ok(Value::String(value_to_string(&value))),
            }
        }
        "replace" => {
            let from = args.first().map(value_to_string).unwrap_or_default();
            let to = args.get(1).map(value_to_string).unwrap_or_default();
            Ok(Value::String(value_to_string(&value).replace(&from, &to)))
        }
        other => Err(IrError::Render(format!("unknown filter `{other}`"))),
    }
}

fn is_truthy(v: &Value) -> bool {
    match v {
        Value::Null => false,
        Value::Bool(b) => *b,
        Value::Number(n) => n.as_f64().map(|f| f != 0.0).unwrap_or(false),
        Value::String(s) => !s.is_empty(),
        Value::Array(a) => !a.is_empty(),
        Value::Object(o) => !o.is_empty(),
    }
}

fn value_to_string(v: &Value) -> String {
    match v {
        Value::Null => String::new(),
        Value::Bool(b) => b.to_string(),
        Value::Number(n) => n.to_string(),
        Value::String(s) => s.clone(),
        other => other.to_string(),
    }
}

fn html_escape(s: &str) -> String {
    s.chars()
        .map(|c| match c {
            '&' => "&amp;".to_string(),
            '<' => "&lt;".to_string(),
            '>' => "&gt;".to_string(),
            '"' => "&quot;".to_string(),
            '\'' => "&#39;".to_string(),
            c => c.to_string(),
        })
        .collect()
}
